chunk_test.go

```diff
@@ -0,0 +1,145 @@
+package gosnowflake
+
+import (
+	"bytes"
+	"encoding/json"
+	"strings"
+	"testing"
+)
+
+func TestBadChunkData(t *testing.T) {
+	testDecodeErr(t, "")
+	testDecodeErr(t, "null")
+	testDecodeErr(t, "42")
+	testDecodeErr(t, "\"null\"")
+	testDecodeErr(t, "{}")
+
+	testDecodeErr(t, "[[]")
+	testDecodeErr(t, "[null]")
+	testDecodeErr(t, `[[hello world]]`)
+
+	testDecodeErr(t, `[[""hello world""]]`)
+	testDecodeErr(t, `[["\"hello world""]]`)
+	testDecodeErr(t, `[[""hello world\""]]`)
+	testDecodeErr(t, `[["hello world`)
+	testDecodeErr(t, `[["hello world"`)
+	testDecodeErr(t, `[["hello world"]`)
+
+	testDecodeErr(t, `[["\uQQQQ"]]`)
+
+	for b := byte(0); b < ' '; b++ {
+		testDecodeErr(t, string([]byte{
+			'[', '[', '"', b, '"', ']', ']',
+		}))
+	}
+}
+
+func TestValidChunkData(t *testing.T) {
+	testDecodeOk(t, "[]")
+	testDecodeOk(t, "[  ]")
+	testDecodeOk(t, "[[]]")
+	testDecodeOk(t, "[ [  ]   ]")
+	testDecodeOk(t, "[[],[],[],[]]")
+	testDecodeOk(t, "[[] , []  , [], []  ]")
+
+	testDecodeOk(t, "[[null]]")
+	testDecodeOk(t, "[[\n\t\r null]]")
+	testDecodeOk(t, "[[null,null]]")
+	testDecodeOk(t, "[[ null , null ]]")
+	testDecodeOk(t, "[[null],[null],[null]]")
+	testDecodeOk(t, "[[null],[ null  ] ,  [null]]")
+
+	testDecodeOk(t, `[[""]]`)
+	testDecodeOk(t, `[["false"]]`)
+	testDecodeOk(t, `[["true"]]`)
+	testDecodeOk(t, `[["42"]]`)
+
+	testDecodeOk(t, `[[""]]`)
+	testDecodeOk(t, `[["hello"]]`)
+	testDecodeOk(t, `[["hello world"]]`)
+
+	testDecodeOk(t, `[["/ ' \\ \b \t \n \f \r \""]]`)
+	testDecodeOk(t, `[["❄"]]`)
+	testDecodeOk(t, `[["\u2744"]]`)
+	testDecodeOk(t, `[["\uFfFc"]]`)       // consume replacement chars
+	testDecodeOk(t, `[["\ufffd"]]`)       // consume replacement chars
+	testDecodeOk(t, `[["\u0000"]]`)       // yes, this is valid
+	testDecodeOk(t, `[["\uD834\uDD1E"]]`) // surrogate pair
+	testDecodeOk(t, `[["\uD834\u0000"]]`) // corrupt surrogate pair
+
+	testDecodeOk(t, `[["$"]]`)      // "$"
+	testDecodeOk(t, `[["\u0024"]]`) // "$"
+
+	testDecodeOk(t, `[["\uC2A2"]]`) // "¢"
+	testDecodeOk(t, `[["¢"]]`)      // "¢"
+
+	testDecodeOk(t, `[["\u00E2\u82AC"]]`) // "€"
```

Comment:
This is not a euro sign but two characters, right?
\u is used to escape UTF-16 code point.

Reply:
It's actually the multi-byte utf-8 encoding for the euro sign. I pull this from the `wikipedia` article (along with the other currency tests).

https://en.wikipedia.org/wiki/UTF-8#Example

Reply:
ok, that's what I thought, but JSON doesn't accept escaped UTF-8 encoding, and `\u00E2\u82AC` won't be converted to \xe2\x82\xac in bytes.

Reply:
According to the RFC, a `UTF-8` encoding is perfectly acceptable. 
> https://tools.ietf.org/html/rfc7159#section-8.1

They also explicitly call out the possibility of a "12-byte" unicode sequence in Section 7.
> To escape an extended character that is not in the Basic Multilingual Plane, the character is represented as a 12-character sequence, encoding the UTF-16 surrogate pair.  So, for example, a string containing only the G clef character (U+1D11E) may be represented as "\uD834\uDD1E".